EIPS/eip-2537.md

```diff
@@ -0,0 +1,296 @@
+
+##### ABI for G1 multiexponentiation
+
+G1 multiplication call expects `160*k` bytes as an input that is interpreted as byte concatenation of `k` slices each of them being a byte concatenation of encoding of G1 point (`128` bytes) and encoding of a scalar value (`32` bytes). Output is an encoding of multiexponentiation operation result - single G1 point (`128` bytes).
+
+Error cases:
+- Any of G1 points being not on the curve must result in error
+- Field elements encoding rules apply (obviously)
+- Input has invalid length
+
+##### ABI for G2 addition
+
+G2 addition call expects `512` bytes as an input that is interpreted as byte concatenation of two G2 points (`256` bytes each). Output is an encoding of addition operation result - single G2 point (`256` bytes).
+
+Error cases:
+- Either of points being not on the curve must result in error
+- Field elements encoding rules apply (obviously)
+- Input has invalid length
+
+##### ABI for G2 multiplication
+
+G2 multiplication call expects `288` bytes as an input that is interpreted as byte concatenation of encoding of G2 point (`256` bytes) and encoding of a scalar value (`32` bytes). Output is an encoding of multiplication operation result - single G2 point (`256` bytes).
+
+Error cases:
+- Point being not on the curve must result in error
+- Field elements encoding rules apply (obviously)
+- Input has invalid length
+
+##### ABI for G2 multiexponentiation
+
+G2 multiplication call expects `288*k` bytes as an input that is interpreted as byte concatenation of `k` slices each of them being a byte concatenation of encoding of G2 point (`256` bytes) and encoding of a scalar value (`32` bytes). Output is an encoding of multiexponentiation operation result - single G2 point (`256` bytes).
+
+Error cases:
+- Any of G2 points being not on the curve must result in error
+- Field elements encoding rules apply (obviously)
+- Input has invalid length
+
+##### ABI for pairing
+
+Pairing call expects `384*k` bytes as an inputs that is interpreted as byte concatenation of `k` slices. Each slice has the following structure:
+- `128` bytes of G1 point encoding
+- `256` bytes of G2 point encoding
+
+Output is a `32` bytes where first `31` bytes are equal to `0x00` and the last byte is `0x01` if pairing result is equal to multiplicative identity in a pairing target field and `0x00` otherwise.
+
+Error cases:
+- Invalid encoding of any boolean variable must result in error
+- Any of G1 or G2 points being not on the curve must result in error
+- Any of G1 or G2 points are not in the correct subgroup
+- Field elements encoding rules apply (obviously)
+- Input has invalid length
+
+##### ABI for mapping field-to-curve
+
+Field-to-curve call expects either `64` or `128` bytes an an input that is interpreted as a an element of the base field or a quadratic extension respectively. Output of this call is either `128` or `256` bytes and is a either G1 or G2 point form following respective encoding rules.
```

Comment:
I'm concerned about the input sizes and output sizes being what appears to be arbirarily different.  If I always pass in field elements (length 64) do I always get out a 128 bytes result?  Same for 128 bytes input and 256 bytes output.  If this is the case perhaps this should be two methods?  The API type checking would be way easier.

Reply:
having two different precompiles is not off the table but does consume more precompile opcodes (a relatively scarce resource, unless we start doing extension tables or something haha)

the setting here is that you have two groups (`G1` and `G2`) taking elements from underlying fields (`Fp` and `Fp2`), respectively -- you *don't* have `Fp2` elements under `G1` and vice versa. `Fp2` here is a "quadratic extension" of `Fp` (just a finite field of some prime order `p`) like the complex numbers are a quadratic extension of the real numbers. So `Fp2` can be represented as `a +ib` for `a,b` in `Fp` -- equivalently packed as just the two `Fp` elements `a` and `b`. So, you are correct that the `64` input maps to the `128` output and the `128` input maps to the `256` output exclusively.

Reply:
wrt to type checking it comes down to having two routines to do two precompiles, or one precompile that has a union type of both field elements (and likely dispatches to two different functions anyway...)

Reply:
Union types lead to surprising bugs.  It should be two precompiles.

Reply:
Also, I don't see precompile space as that relatively scarce.  We have a large amount of space before we run into our first hash collision.